### ISSUE TITLE
Visual design review of combobox PR

### ISSUE THREAD

**Issue by zcorpan** [w3c/aria-practices#1263]
[Visual design review](https://github.com/w3c/aria-practices/wiki/Pull-Request-Review-Process#design_review) of combobox PR https://github.com/w3c/aria-practices/pull/1255

> Review the visual presentation to assess whether look and feel is similar to typical modern implementations of the widget. Allow for adjustments to accommodate accessibility requirements in cases where most modern implementations are not sufficiently accessible.

There are 4 examples:

* [Editable Combobox With Both List and Inline Autocomplete Example](https://raw.githack.com/w3c/aria-practices/issue1244-aria1.2-combobox-conversion/examples/combobox/combobox-autocomplete-both.html)
* [Editable Combobox With List Autocomplete Example](https://raw.githack.com/w3c/aria-practices/issue1244-aria1.2-combobox-conversion/examples/combobox/combobox-autocomplete-list.html)
* [Editable Combobox without Autocomplete Example](https://raw.githack.com/w3c/aria-practices/issue1244-aria1.2-combobox-conversion/examples/combobox/combobox-autocomplete-none.html)
* [Editable Combobox with Grid Popup Example](https://raw.githack.com/w3c/aria-practices/issue1244-aria1.2-combobox-conversion/examples/combobox/grid-combo.html)

The first 3 have the same visual design, so I'll only review the first one of those ("example 1"). The grid popup example ("example 4") has different visual design.

## Example 1

First, let's compare the default state of the example side by side in Safari (left), Chrome (middle), and Firefox (right).

<img width="742" alt="The borders are different. The spacing between the combobox and the button is different. The combobox and the button don't align in Safari and Chrome." src="https://user-images.githubusercontent.com/244772/68805565-5cb3b980-0664-11ea-97bf-3483454d3d91.png">

The following screenshots are taken with Firefox Nightly on macOS.

Default state
<img width="365" alt="The label State on one line, a gray textfield with a 2px inset border, next to it a gray button with a 2px outset border and a downward-pointing triangle." src="https://user-images.githubusercontent.com/244772/68807368-fd57a880-0667-11ea-89f0-a1d20223995c.png">

Hover on the combobox field
<img width="377" alt="The combobox has a black outline." src="https://user-images.githubusercontent.com/244772/68807428-106a7880-0668-11ea-96e5-1fed21d66e05.png">

Hover on the button
<img width="358" alt="The button has a black outline." src="https://user-images.githubusercontent.com/244772/68807465-1d876780-0668-11ea-97b3-88e43e631324.png">

Focus on the combobox
<img width="340" alt="The combobox has a blue outline and a light blue background." src="https://user-images.githubusercontent.com/244772/68805890-009d6500-0665-11ea-943f-4bd050d5d106.png">

Popup open
<img width="333" alt="The popup is shown underneath the combobox. It has a gray background and a black border. The popup overlaps other content. The popup is wider than the combobox field, but narrower than the combobox and the button together." src="https://user-images.githubusercontent.com/244772/68805963-232f7e00-0665-11ea-966a-b7fba5ea7fa7.png">

After typing "Al" and focusing Alabama
<img width="334" alt="The combobox has a black outline and contains the text Alabama. The popup has a blue border and shows the options Alabama and Alaska. Alabama in the popup has a light blue background. The popup has the same height as before." src="https://user-images.githubusercontent.com/244772/68806177-9df89900-0665-11ea-8307-7ee0b936eb5e.png">

## Example 4

Default state in Safari (top), Chrome (middle), Firefox (bottom)
<img width="450" alt="The combobox looks the same across browsers. It looks like a white text box with a thin gray border." src="https://user-images.githubusercontent.com/244772/68808634-735d0f00-066a-11ea-84b0-efd629a03ec9.png">

There is no provided styling for hover or focus. In Safari and Chrome, the default focus ring is shown. In Firefox, it is not. Removing the `border-radius` makes Firefox show its default focus ring. This might be a bug in Firefox? (Specifying `-webkit-appearance: none` has the same effect.)

After focusing and typing the letter "a"
<img width="421" alt="A white popup with thin gray border is shown underneath the combobox, with options Apple, Artichoke, Asparagus, and each option has smaller gray text on the right hand side - Fruit, Vegetable, Vegetable. The popup is almost as wide as the combobox" src="https://user-images.githubusercontent.com/244772/68808870-f1b9b100-066a-11ea-9ebc-6d2d07ec6441.png">

Moving focus to the first cell of the first row
<img width="402" alt="The first row has a blue background. The text Apple has a green dotted border around it." src="https://user-images.githubusercontent.com/244772/68808876-f8482880-066a-11ea-8250-89d8f380bf93.png">

Moving focus to the second cell of the first row
<img width="401" alt="Same as previous except the green dotted border is around Fruit." src="https://user-images.githubusercontent.com/244772/68808884-fd0cdc80-066a-11ea-907c-b87a4ea1fa79.png">

## Typical modern implementations of the widget

* [Material design exposed dropdown menu](https://material.io/components/menus/menus.html#exposed-dropdown-menu)
  - There are variations, but it label (if any) appears within the widget, and the dropdown button is also within the widget. The popup usually has the same width as the combobox.
* [PolymerElements `paper-dropdown-menu` demo](https://www.webcomponents.org/element/@polymer/paper-dropdown-menu/demo/demo/index.html)
  - I think this implements the previous version of Material Design. This demo doesn't include an editable combobox.
* [Bootstrap has dropdown buttons](https://getbootstrap.com/docs/4.0/components/dropdowns/) but not comboboxes?
* [Vaadin Combo Box](https://vaadin.com/components/vaadin-combo-box)
* [PatternFly Elements `pfe-autocomplete`](https://patternfly.github.io/patternfly-elements/demo/?path=/story/autocomplete--pfe-autocomplete)
* [BluePrint `Suggest`](https://blueprintjs.com/docs/#select/suggest)
* [BluePrint `Omnibar`](https://blueprintjs.com/docs/#select/omnibar)
  - The popup here has years on the right, similar to Example 4. But it doesn't allow selecting individual cells with left and right arrows.

## Analysis

Our examples are not on par with modern design systems/component frameworks.

Example 1:

- It is inconsistent between browsers.
- It is, in my opinion, ugly. It doesn't look modern at all.
- The black outline on hover is too stark and isn't found in other implementations. If anything, a hover effect is usually a subtle change in background color.
- The dropdown button should be inside the combobox.
- The popup should have the same width as the combobox.
- The popup should have a nicer looking border than the thick black border.
- The border style should be changed to look more modern.

Example 4:

- In its default state, it looks reasonably modern, like a normal text field.
- No custom focus style; in Firefox the applied styling but lack of focus style results in no focus indication (other than blinking cursor).
- No indication that it is a combobox; maybe there should be a dropdown button?
- The focus style in the popup (dotted green border) is ugly.
- The ability to focus individual cells is confusing to me - selecting "Fruit" (second cell) results in "Apples" (first cell) being used as the value. Is this a good example for a Grid popup at all? (Maybe this is more of a functional review than design review, though.)

**Comment by JAWS-test** [comment]
> No indication that it is a combobox; maybe there should be a dropdown button?

I think it is very important to have a visual indication for the role. A combobox should be visually different from an input field. A combobox with text input should also be visually distinguishable from a combobox without text input. Just like a `<select size=1>` should differ from an input field with `<datalist>`, because the handling is different.

**Comment by JAWS-test** [comment]
The focus indicator shows insufficient contrasts and thus violates WCAG SC 1.4.11:
* blue background to white background: 2.0:1
* green frame on blue background: 2.6:1

Minimum contrast is 3:1.

In addition, WCAG 2.2 requires a minimum area for the contrast indicator (length of the element times 2px). I am unsure whether the green frame meets this requirement.

**Comment by mcking65** [comment]
Per review by @smhigley in #1255: The selected option within the listbox is not visible in Windows high contrast mode.

**Comment by zcorpan** [comment]
@carmacleod wrote

> the black hover outline appears when hovering over what looks like empty space off to the right of the button (but is actually the combo's label element). This feels unexpected.

This is because the `label` element is `display: block`. It's not just unexpected hover style; clicking or tapping there, with the intention of moving focus *away* from the combobox, has the opposite effect.

**Comment by carmacleod** [comment]
Wow @zcorpan ! Your commit https://github.com/w3c/aria-practices/pull/1255/commits/73ca367d89785f923ccc557d386916a90896fc7e makes the listbox combos look WAY better!

There's a Windows-Chrome issue with the [2nd](https://raw.githack.com/w3c/aria-practices/issue1244-aria1.2-combobox-conversion/examples/combobox/combobox-autocomplete-list.html) and [3rd](https://raw.githack.com/w3c/aria-practices/issue1244-aria1.2-combobox-conversion/examples/combobox/combobox-autocomplete-none.html) listbox combo, where the popup is skewed to the left edge of the label. Screenshots below:
![image](https://user-images.githubusercontent.com/3331913/68872211-a84d7e00-06cb-11ea-9149-20b2d4aad76c.png)

![image](https://user-images.githubusercontent.com/3331913/68872294-cca95a80-06cb-11ea-84c8-8f7618b773dc.png)

**Comment by zcorpan** [comment]
I have made changes to examples 1-3 to try to address the design issues.

https://rawgit.com/w3c/aria-practices/issue1244-aria1.2-combobox-conversion/examples/combobox/combobox-autocomplete-both.html

It should work with Windows High Contrast Mode in Edge as well.

![Screenshot showing dropdown option using highlight color in Windows High Contrast Mode.](https://user-images.githubusercontent.com/244772/68875062-7d325100-0702-11ea-8406-cd38d02a3084.png)

I haven't changed example 4.

**Comment by carmacleod** [comment]
I liked the rounder corners... :)

**Comment by zcorpan** [comment]
Yeah though @isaacdurazo said rounded corners are a bit passe. 🙂

**Comment by isaacdurazo** [comment]
I also think the usage of rounded corners as well as some colors were inconsistent with the design of the rest of the page

**Comment by zcorpan** [comment]
Thanks @carmacleod! The labels in 2 and 3 should be moved outside of the combobox container in the markup.

**Comment by carmacleod** [comment]
>  @isaacdurazo said rounded corners are a bit passe. 🙂

Trends come and go... but science is here to stay. 😉 
https://uxmovement.com/thinking/why-rounded-corners-are-easier-on-the-eyes/
